File: ForkandBeard/AutoBalancingFormTimer.vb
#Region " Options "
Option Explicit On
Option Strict On
#End Region

Public Class AutoBalancingFormTimer : Inherits Windows.Forms.Timer

    Private _TockLongerThanMaxIntervalOccured As Boolean
    Private _MinInterval As Integer = -1
    Private _MaxInterval As Integer = -1
    Event BalancedTock()
    Event TockLongerThanMaxInterval(ByVal pdblDuration As Double)
    Event AverageTockIntervalResumed()

    Public Property MinInterval() As Integer
        Get
            Return Me._MinInterval
        End Get
        Set(ByVal value As Integer)
            Me._MinInterval = value
        End Set
    End Property

    Public Property MaxInterval() As Integer
        Get
            Return Me._MaxInterval
        End Get
        Set(ByVal value As Integer)
            Me._MaxInterval = value
        End Set
    End Property

    Private Sub AutoBalancingFormTimer_Tick(ByVal sender As Object, ByVal e As System.EventArgs) Handles Me.Tick
        Dim dteStart As Date
        Dim dblDuration As Double
        dteStart = Date.Now

        RaiseEvent BalancedTock()
        dblDuration = Date.Now.Subtract(dteStart).TotalMilliseconds()

        If Me.MinInterval <> -1 AndAlso Me.MaxInterval <> -1 Then

            If dblDuration > Me.MaxInterval Then
                If Not Me._TockLongerThanMaxIntervalOccured Then
                    Me._TockLongerThanMaxIntervalOccured = True
                    RaiseEvent TockLongerThanMaxInterval(dblDuration)
                End If
            End If

            If (dblDuration * 1.1) > Me.Interval Then
                Me.Interval = CInt(Math.Min(dblDuration * 1.2, Me.MaxInterval))
            ElseIf Me.Interval <> Me.MinInterval Then
                If dblDuration * 1.1 <= Me.Interval Then
                    Me.Interval = CInt(Math.Max(Me.Interval * 0.8, Me.MinInterval))

                    If Me._TockLongerThanMaxIntervalOccured _
                    AndAlso Me.Interval <= (Me.MaxInterval - Me.MinInterval) Then
                        Me._TockLongerThanMaxIntervalOccured = False
                        RaiseEvent AverageTockIntervalResumed()
                    End If
                End If
            End If
        End If
    End Sub

    Public Sub New(ByVal pobjContainer As System.ComponentModel.IContainer)
        MyBase.New(pobjContainer)
    End Sub
End Class

File: ForkandBeard.Util/UI/AutoBalancingFormTimer.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;

namespace ForkandBeard.Util.UI
{
    public partial class AutoBalancingFormTimer : System.Windows.Forms.Timer
    {
        private bool _TockLongerThanMaxIntervalOccured;
        private int _MinInterval = -1;
        private int _MaxInterval = -1;
        public event BalancedTockEventHandler BalancedTock;
        public delegate void BalancedTockEventHandler();
        public event TockLongerThanMaxIntervalEventHandler TockLongerThanMaxInterval;
        public delegate void TockLongerThanMaxIntervalEventHandler(double pdblDuration);
        public event AverageTockIntervalResumedEventHandler AverageTockIntervalResumed;
        public delegate void AverageTockIntervalResumedEventHandler();

        public int MinInterval
        {
            get { return this._MinInterval; }
            set { this._MinInterval = value; }
        }

        public int MaxInterval
        {
            get { return this._MaxInterval; }
            set { this._MaxInterval = value; }
        }

        private void AutoBalancingFormTimer_Tick(object sender, System.EventArgs e)
        {
            System.DateTime dteStart = default(System.DateTime);
            double dblDuration = 0;
            dteStart = System.DateTime.Now;

            if (BalancedTock != null)
            {
                BalancedTock();
            }
            dblDuration = System.DateTime.Now.Subtract(dteStart).TotalMilliseconds;


            if (this.MinInterval != -1 && this.MaxInterval != -1)
            {
                if (dblDuration > this.MaxInterval)
                {
                    if (!this._TockLongerThanMaxIntervalOccured)
                    {
                        this._TockLongerThanMaxIntervalOccured = true;
                        if (TockLongerThanMaxInterval != null)
                        {
                            TockLongerThanMaxInterval(dblDuration);
                        }
                    }
                }

                if ((dblDuration * 1.1) > this.Interval)
                {
                    this.Interval = Convert.ToInt32(Math.Min(dblDuration * 1.2, this.MaxInterval));
                }
                else if (this.Interval != this.MinInterval)
                {
                    if (dblDuration * 1.1 <= this.Interval)
                    {
                        this.Interval = Convert.ToInt32(Math.Max(this.Interval * 0.8, this.MinInterval));

                        if (this._TockLongerThanMaxIntervalOccured && this.Interval <= (this.MaxInterval - this.MinInterval))
                        {
                            this._TockLongerThanMaxIntervalOccured = false;
                            if (AverageTockIntervalResumed != null)
                            {
                                AverageTockIntervalResumed();
                            }
                        }
                    }
                }
            }
        }

        public AutoBalancingFormTimer(System.ComponentModel.IContainer pobjContainer)
            : base(pobjContainer)
        {
            Tick += AutoBalancingFormTimer_Tick;
        }
    }
}